protocol/flows/relaytransactions/relaytransactions.go

```diff
@@ -0,0 +1,227 @@
+package relaytransactions
+
+import (
+	"github.com/kaspanet/kaspad/blockdag"
+	"github.com/kaspanet/kaspad/mempool"
+	"github.com/kaspanet/kaspad/netadapter"
+	"github.com/kaspanet/kaspad/netadapter/router"
+	peerpkg "github.com/kaspanet/kaspad/protocol/peer"
+	"github.com/kaspanet/kaspad/protocol/protocolerrors"
+	"github.com/kaspanet/kaspad/util"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/wire"
+	"github.com/pkg/errors"
+	"time"
+)
+
+const timeout = 30 * time.Second
+
+// NewBlockHandler is a function that is to be
+// called when a new block is successfully processed.
+type NewBlockHandler func(block *util.Block) error
+
+// HandleRelayedTransactions listens to wire.MsgTxInv messages, requests their corresponding transactions if they
+// are missing, adds them to the mempool and propagates them to the rest of the network.
+func HandleRelayedTransactions(incomingRoute *router.Route, outgoingRoute *router.Route,
+	netAdapter *netadapter.NetAdapter, dag *blockdag.BlockDAG, txPool *mempool.TxPool,
+	sharedRequestedTransactions *SharedRequestedTransactions) error {
+
+	invsQueue := make([]*wire.MsgTxInv, 0)
+	for {
+		inv, shouldStop, err := readInv(incomingRoute, &invsQueue)
+		if err != nil {
+			return err
+		}
+		if shouldStop {
+			return nil
+		}
+
+		requestedIDs, shouldStop, err := requestInvTransactions(outgoingRoute, txPool, dag, sharedRequestedTransactions, inv)
+		if err != nil {
+			return err
+		}
+		if shouldStop {
+			return nil
+		}
+
+		shouldStop, err = receiveTransactions(requestedIDs, incomingRoute, &invsQueue, txPool, netAdapter,
+			sharedRequestedTransactions)
+		if err != nil {
+			return err
+		}
+		if shouldStop {
+			return nil
+		}
+	}
+}
+
+func requestInvTransactions(outgoingRoute *router.Route, txPool *mempool.TxPool, dag *blockdag.BlockDAG,
+	sharedRequestedTransactions *SharedRequestedTransactions, inv *wire.MsgTxInv) (requestedIDs []*daghash.TxID,
+	shouldStop bool, err error) {
+
+	idsToRequest := make([]*daghash.TxID, 0, len(inv.TXIDs))
```

Comment:
Why the whole jumping-through-hoops with idsToRequest? I think updateQueues could return it